crates/ra_hir/src/expr/lower.rs

```diff
@@ -1,593 +0,0 @@
-        self.body.pats.alloc(Pat::Missing)
-    }
-
-    fn collect_expr(&mut self, expr: ast::Expr) -> ExprId {
-        let syntax_ptr = AstPtr::new(&expr);
-        match expr {
-            ast::Expr::IfExpr(e) => {
-                let then_branch = self.collect_block_opt(e.then_branch());
-
-                let else_branch = e.else_branch().map(|b| match b {
-                    ast::ElseBranch::Block(it) => self.collect_block(it),
-                    ast::ElseBranch::IfExpr(elif) => {
-                        let expr: ast::Expr = ast::Expr::cast(elif.syntax().clone()).unwrap();
-                        self.collect_expr(expr)
-                    }
-                });
-
-                let condition = match e.condition() {
-                    None => self.missing_expr(),
-                    Some(condition) => match condition.pat() {
-                        None => self.collect_expr_opt(condition.expr()),
-                        // if let -- desugar to match
-                        Some(pat) => {
-                            let pat = self.collect_pat(pat);
-                            let match_expr = self.collect_expr_opt(condition.expr());
-                            let placeholder_pat = self.missing_pat();
-                            let arms = vec![
-                                MatchArm { pats: vec![pat], expr: then_branch, guard: None },
-                                MatchArm {
-                                    pats: vec![placeholder_pat],
-                                    expr: else_branch.unwrap_or_else(|| self.empty_block()),
-                                    guard: None,
-                                },
-                            ];
-                            return self
-                                .alloc_expr(Expr::Match { expr: match_expr, arms }, syntax_ptr);
-                        }
-                    },
-                };
-
-                self.alloc_expr(Expr::If { condition, then_branch, else_branch }, syntax_ptr)
-            }
-            ast::Expr::TryBlockExpr(e) => {
-                let body = self.collect_block_opt(e.body());
-                self.alloc_expr(Expr::TryBlock { body }, syntax_ptr)
-            }
-            ast::Expr::BlockExpr(e) => self.collect_block(e),
-            ast::Expr::LoopExpr(e) => {
-                let body = self.collect_block_opt(e.loop_body());
-                self.alloc_expr(Expr::Loop { body }, syntax_ptr)
-            }
-            ast::Expr::WhileExpr(e) => {
-                let body = self.collect_block_opt(e.loop_body());
-
-                let condition = match e.condition() {
-                    None => self.missing_expr(),
-                    Some(condition) => match condition.pat() {
-                        None => self.collect_expr_opt(condition.expr()),
-                        // if let -- desugar to match
-                        Some(pat) => {
-                            tested_by!(infer_while_let);
```

Comment:
Note to self:

this used type-inference to test lowering,  which is a bit problematic as far as dependencies go. 
Perhaps we ca re-cast this test as a test for name resolution? It should then be handled by `hir_def` crate.